File: Observer/NotificationServiceObserver.cs
using System;
using System.Collections.Generic;

// Interfaz para los observadores
public interface IObserver {
    void Update(string message);
}

// Clase concreta que implementa el observador
public class User : IObserver {
    private string name;

    public User(string name) {
        this.name = name;
    }

    public void Update(string message) {
        Console.WriteLine($"{name} received: {message}");
    }
}

// Clase que implementa el sujeto observado
public class NotificationService {
    private List<IObserver> observers = new List<IObserver>();

    // Método para suscribir observadores
    public void Subscribe(IObserver observer) {
        observers.Add(observer);
    }

    // Método para notificar a los observadores
    public void NotifyObservers(string message) {
        foreach (var observer in observers) {
            observer.Update(message);
        }
    }
}

// Uso del patrón Observer
class Program {
    static void Main(string[] args) {
        NotificationService service = new NotificationService();
        service.Subscribe(new User("Alice"));
        service.Subscribe(new User("Bob"));

        service.NotifyObservers("New update available!");
    }
}

File: Observer/notification_service_observer.py
# Clase base para observadores
class Observer:
    def update(self, message):
        pass

# Clase concreta que implementa el observador
class User(Observer):
    def __init__(self, name):
        self.name = name

    def update(self, message):
        print(f"{self.name} received: {message}")

# Clase que implementa el sujeto observado
class NotificationService:
    def __init__(self):
        self.observers = []

    # Método para suscribir observadores
    def subscribe(self, observer):
        self.observers.append(observer)

    # Método para notificar a los observadores
    def notify_observers(self, message):
        for observer in self.observers:
            observer.update(message)

# Uso del patrón Observer
service = NotificationService()
service.subscribe(User("Alice"))
service.subscribe(User("Bob"))

service.notify_observers("New update available!")
